File: streampark-spark/streampark-spark-core/src/main/scala/org/apache/streampark/spark/core/serializable/SerializationExtensions.scala
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.spark.core.serializable

import org.apache.avro.Schema
import org.apache.avro.generic.GenericRecord
import org.apache.avro.mapred.AvroKey
import org.apache.avro.mapreduce.{AvroJob, AvroKeyInputFormat, AvroKeyOutputFormat}
import org.apache.avro.specific.SpecificData
import org.apache.hadoop.conf.Configuration
import org.apache.hadoop.io.NullWritable
import org.apache.hadoop.mapreduce.Job
import org.apache.log4j.Logger
import org.apache.spark.SparkContext
import org.apache.spark.rdd.RDD

import scala.reflect.{classTag, ClassTag}

object SerializationExtensions {

  private val logger = Logger.getLogger(getClass)

  /**
   * Build a AvroJob
   *
   * @param job
   * @tparam T
   * @return
   */
  def avroJob[T <: GenericRecord: ClassTag](
      job: Job = Job.getInstance(new Configuration())): Job = {
    val schema: Schema = SpecificData.get.getSchema(classTag[T].runtimeClass)
    AvroJob.setInputKeySchema(job, schema)
    AvroJob.setOutputKeySchema(job, schema)
    job
  }

  def isDefined(record: GenericRecord, field: String): Boolean = {
    if (record.get(field) != null) return true
    logger.warn(
      s"Expected field '$field' to be defined, but it was not on record of type '${record.getClass}'")
    false
  }

  implicit class AvroRDDSparkContext(val sparkContext: SparkContext) extends AnyVal {

    /**
     * Read the Avro file in the specified path
     *
     * @param path
     * @tparam T
     * @return
     */
    def avroFile[T: ClassTag](path: String): RDD[T] = {
      sparkContext
        .newAPIHadoopFile(
          path,
          classOf[AvroKeyInputFormat[T]],
          classOf[AvroKey[T]],
          classOf[NullWritable],
          avroJob().getConfiguration)
        .map[T](_._1.datum())
    }
  }

  implicit class AvroRDD[T <: GenericRecord: ClassTag](val avroRDD: RDD[T]) {
    def filterIfUnexpectedNull(fields: String*): RDD[T] = {
      avroRDD.filter(r => fields.forall(isDefined(r, _)))
    }

    /** Save as avro file */
    def saveAsAvroFile(outputPath: String): Unit = {
      avroRDD
        .map(r => (new AvroKey[T](r), NullWritable.get))
        .saveAsNewAPIHadoopFile(
          outputPath,
          classOf[AvroKey[T]],
          classOf[NullWritable],
          classOf[AvroKeyOutputFormat[T]],
          avroJob[T]().getConfiguration)
    }

    /**
     * Save the specified avro files according to the Key value
     *
     * @param outputKeyFun
     *   outputKeyFunction
     * @param outputPath
     *   outputPath
     */
    def saveAsMultipleAvroFiles(outputKeyFun: (T) => String, outputPath: String): Unit = {
      avroRDD
        .map(r => ((outputKeyFun(r), new AvroKey(r)), NullWritable.get))
        .saveAsNewAPIHadoopFile(
          outputPath,
          classOf[AvroKey[(String, T)]],
          classOf[NullWritable],
          classOf[MultipleAvroOutputsFormat[T]],
          avroJob[T]().getConfiguration)
    }
  }

}

File: streampark-spark/streampark-spark-core/src/main/java/org/apache/streampark/spark/core/serializable/SerializationExtensions.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.streampark.spark.core.serializable;

import java.io.IOException;

import org.apache.avro.Schema;
import org.apache.avro.generic.GenericRecord;
import org.apache.avro.mapred.AvroKey;
import org.apache.avro.mapreduce.AvroJob;
import org.apache.avro.mapreduce.AvroKeyInputFormat;
import org.apache.avro.mapreduce.AvroKeyOutputFormat;
import org.apache.avro.specific.SpecificData;
import org.apache.hadoop.conf.Configuration;
import org.apache.hadoop.io.NullWritable;
import org.apache.hadoop.mapreduce.Job;
import org.apache.log4j.Logger;
import org.apache.spark.SparkContext;
import org.apache.spark.api.java.JavaPairRDD;
import org.apache.spark.api.java.JavaRDD;

/** Avro serialization helpers for Spark. */
public final class SerializationExtensions {

    private static final Logger LOGGER = Logger.getLogger(SerializationExtensions.class);

    private SerializationExtensions() {}

    public static <T extends GenericRecord> Job avroJob(Class<T> clazz) {
        try {
            return avroJob(clazz, Job.getInstance(new Configuration()));
        } catch (IOException e) {
            throw new RuntimeException(e);
        }
    }

    public static <T extends GenericRecord> Job avroJob(Class<T> clazz, Job job) {
        Schema schema = SpecificData.get().getSchema(clazz);
        AvroJob.setInputKeySchema(job, schema);
        AvroJob.setOutputKeySchema(job, schema);
        return job;
    }

    public static boolean isDefined(GenericRecord record, String field) {
        if (record.get(field) != null) {
            return true;
        }
        LOGGER.warn(
                "Expected field '"
                        + field
                        + "' to be defined, but it was not on record of type '"
                        + record.getClass()
                        + "'");
        return false;
    }

    public static <T extends GenericRecord> JavaRDD<T> avroFile(
            SparkContext sparkContext, String path, Class<T> clazz) {
        Job job = avroJob(clazz);
        JavaPairRDD<AvroKey, NullWritable> pairRDD =
                JavaPairRDD.fromJavaRDD(
                        sparkContext
                                .newAPIHadoopFile(
                                        path,
                                        AvroKeyInputFormat.class,
                                        AvroKey.class,
                                        NullWritable.class,
                                        job.getConfiguration())
                                .toJavaRDD());
        return pairRDD.map(tuple -> (T) tuple._1.datum());
    }

    public static <T extends GenericRecord> JavaRDD<T> filterIfUnexpectedNull(
            JavaRDD<T> avroRDD, String... fields) {
        return avroRDD.filter(
                r -> {
                    for (String field : fields) {
                        if (!isDefined(r, field)) {
                            return false;
                        }
                    }
                    return true;
                });
    }

    public static <T extends GenericRecord> void saveAsAvroFile(
            JavaRDD<T> avroRDD, String outputPath, Class<T> clazz) throws Exception {
        Job job = avroJob(clazz);
        JavaPairRDD<AvroKey<T>, NullWritable> output =
                avroRDD.mapToPair(r -> new scala.Tuple2<>(new AvroKey<>(r), NullWritable.get()));
        output.saveAsNewAPIHadoopFile(
                outputPath,
                AvroKey.class,
                NullWritable.class,
                AvroKeyOutputFormat.class,
                job.getConfiguration());
    }
}
